contracts/src/GaspToken.sol

```diff
@@ -1,30 +1,126 @@
-// SPDX-License-Identifier: MIT
-pragma solidity ^0.8.12;
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.13;
 
-import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {Context} from "@openzeppelin/contracts/utils/Context.sol";
+import {IGaspToken} from "./interfaces/IGaspToken.sol";
 
-contract MintableERC20 is ERC20 {
-    address public owner;
-    constructor(string memory name_, string memory symbol_) ERC20(name_, symbol_) {
-        owner = msg.sender;
-        _mint(msg.sender, 1000000000 * 10 ** decimals());
+contract GaspToken is Context, Ownable, ERC20, IGaspToken {
+    uint256 private constant _TOTAL_SUPPLY = 1_000_000_000 * 10 ** 18;
+    string private constant _NAME = "GASP";
+    string private constant _SYMBOL = "GASP";
+
+    mapping(address => bool) public override whitelist;
+    address public override uniswapPool;
+    bool public override allowTransfers;
+
+    modifier isWhitelisted(address[2] memory addresses, bytes4 selector) {
+        if (!allowTransfers) {
+            if (!whitelist[addresses[0]] && !whitelist[addresses[1]] || addresses[1] == uniswapPool) {
```

Comment:
This seems unclear to me...
Could you please make the order of evaluation explicit?

Reply:
Also, if you are dealling with the two addresses differently (address[0] and address[1]) then I'd rather you use separate function arguments rather than an array

Reply:
Also, why is UniswapPool special?

Reply:
Also, the check seems to be in violation of
"Transfer of tokens, tokens can leave only from addresses included in the "FROM whitelist""

Reply:
You mean the old spec which declared two whitelists earlier. According to our latest discussions, we decided to use a single whitelist both for the sender and the recipient. However, one edge case here should be handled exclusively here. It's a Uniswap pool. The recent spec says that a user can buy tokens from it, but is unable to sell it there. So technically it means that we should explicitly forbid any kind of transfer operation where the Uniswap pool is the recipient.

Reply:
uniswap is special because you can only buy tokens from it so it needs to be the FROM address only whitelist and cannot be inside whitelist that is checking both.

I think the problem that this is a bit unclear is because of address[2] memory addresses usage. I see its used with mixed values, since the value is evaluated the same way, its ok for me.